velox/common/caching/SsdCache.cpp

```diff
@@ -0,0 +1,505 @@
+    int64_t toWrite = 0;
+    for (; begin < pins.size(); ++begin) {
+      auto entry = pins[begin].entry();
+      if (entry->size() > available) {
+        break;
+      }
+      available -= entry->size();
+      toWrite += entry->size();
+    }
+    if (toWrite) {
+      regionSize_[region] += toWrite;
+      return {region * kRegionSize + offset, toWrite};
+    }
+    // A region has been filled. Set its score to be at least the best
+    // score + kRegionSize so that it gets time to live. Otherwise it
+    // has had the least time to get hits and would be the first
+    // evicted.
+    uint64_t best = 0;
+    for (auto& score : regionScore_) {
+      best = std::max<uint64_t>(best, score);
+    }
+    regionScore_[region] = std::max(regionScore_[region], best + kRegionSize);
+    writableRegions_.erase(writableRegions_.begin());
+  }
+}
+
+bool SsdFile::evictLocked() {
+  if (numRegions_ < maxRegions_) {
+    auto newSize = (numRegions_ + 1) * kRegionSize;
+    auto rc = ftruncate(fd_, newSize);
+    if (rc >= 0) {
+      fileSize_ = newSize;
+      writableRegions_.push_back(numRegions_);
+      regionSize_[numRegions_] = 0;
+      ++numRegions_;
+      return true;
+    } else {
+      LOG(ERROR) << "Failed to grow cache file " << filename_ << " to "
+                 << newSize;
+    }
+  }
+  std::vector<int32_t> candidates;
+  int64_t scoreSum = 0;
+  for (int i = 0; i < numRegions_; ++i) {
+    if (regionPins_[i]) {
+      continue;
+    }
+    if (candidates.empty() || regionScore_[i] < scoreSum / candidates.size()) {
+      scoreSum += regionScore_[i];
+      candidates.push_back(i);
+    }
+  }
+  if (candidates.empty()) {
+    suspended_ = true;
+    return false;
+  }
+  std::sort(
+      candidates.begin(), candidates.end(), [&](int32_t left, int32_t right) {
+        return regionScore_[left] < regionScore_[right];
+      });
+  // Free up to 3 lowest score regions.
+  if (candidates.size() > 3) {
+    candidates.resize(3);
+  }
+  clearRegionEntriesLocked(candidates);
+  writableRegions_ = std::move(candidates);
+  suspended_ = false;
+  return true;
+}
+
+void SsdFile::clearRegionEntriesLocked(
+    const std::vector<int32_t>& regionIndices) {
+  // Remove all 'entries_' where the dependent points one of 'regionIndices'.
+  auto it = entries_.begin();
+  while (it != entries_.end()) {
+    auto region = regionIndex(it->second.offset());
+    if (std::find(regionIndices.begin(), regionIndices.end(), region) !=
+        regionIndices.end()) {
+      it = entries_.erase(it);
+    } else {
+      ++it;
+    }
+  }
+  for (auto region : regionIndices) {
+    // While the region is being filled it may get score from
+    // hits. When it is full, it will get a score boost to be a little
+    // ahead of the best.
+    regionScore_[region] = 0;
+    regionSize_[region] = 0;
+  }
+}
+
+void SsdFile::store(std::vector<CachePin>& pins) {
+  std::sort(pins.begin(), pins.end());
```

Comment:
What is the point of sorting here? Do you need regions from the same file to be together?